src/throttler/remote_thottler.js

```diff
@@ -0,0 +1,185 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import Metrics from '../metrics/metrics.js';
+import NullLogger from '../logger.js';
+import NoopMetricFactory from '../metrics/noop/metric_factory';
+import Utils from '../util.js';
+
+const DEFAULT_REFRESH_INTERVAL_MS = 5000;
+const DEFAULT_INITIAL_DELAY_MS = 5000;
+const DEFAULT_THROTTLER_HOST = '0.0.0.0';
+const DEFAULT_THROTTLER_PORT = 5778;
+
+// minimumCredits is the minimum amount of credits necessary to not be throttled.
+// i.e. if currentCredits > MINIMUM_CREDITS, then the operation will not be throttled.
+const MINIMUM_CREDITS = 1.0;
+
+type CreditsPerOperation = { [key: string]: number };
+
+export default class RemoteThrottler {
+  _serviceName: string;
+  _logger: Logger;
+  _metrics: Metrics;
+
+  _refreshIntervalMs: number;
+  _host: string;
+  _port: number;
+
+  _uuid: string;
+  _credits: CreditsPerOperation;
+
+  _initialDelayTimeoutHandle: any;
+  _refreshIntervalHandle: any;
+  _onCreditsUpdate: ?Function;
+
+  /**
+   * Creates a RemoteThrottler that fetches credits remotely from jaeger-agent.
+   *
+   * @param {string} [serviceName] - name of the current service / application, same as given to Tracer
+   * @param {object} [options] - optional settings
+   * @param {object} [options.logger] - optional logger, see _flow/logger.js
+   * @param {object} [options.metrics] - instance of Metrics object
+   * @param {number} [options.refreshIntervalMs] - interval in milliseconds that determines how often credits
+   * are fetched (0 to not refresh, NOT RECOMMENDED!)
+   * @param {number} [options.initialDelayMs] - interval in milliseconds that determines how soon after initialization
+   * credits are first fetched (0 to not refresh, NOT RECOMMENDED!).
+   * @param {string} [options.host] - host for jaeger-agent, defaults to 'localhost'
+   * @param {number} [options.port] - port for jaeger-agent for /credits endpoint
+   * @param {function} [options.onCreditsUpdate] - callback function once credits are updated. Used for testing.
+   */
+  constructor(serviceName: string, options: any = {}) {
+    this._serviceName = serviceName;
+    this._logger = options.logger || new NullLogger();
+    this._metrics = options.metrics || new Metrics(new NoopMetricFactory());
+    this._refreshIntervalMs = options.refreshIntervalMs || DEFAULT_REFRESH_INTERVAL_MS;
+    this._host = options.host || DEFAULT_THROTTLER_HOST;
+    this._port = options.port || DEFAULT_THROTTLER_PORT;
+
+    this._credits = Object.create(null);
```

Comment:
What's the advantage of `Object.create(null)` vs `{}` in this case?

Reply:
I copied over legacy code that did this. Since we're using the keys of the object, won't Object.keys({}) return things like _proto which is something we don't want right? I'll check

Reply:
It should be fine:

>The object of which the enumerable's own properties are to be returned.

[MDN Object.keys](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/keys)